File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/util/FXMLUtil.java
package com.mj.tic.tac.toe.javafx.java.util;

import javafx.fxml.FXMLLoader;
import javafx.scene.layout.Pane;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.io.IOException;
import java.util.Optional;
import java.util.ResourceBundle;
import java.util.stream.Stream;

/**
 * @author Montaser Sbaih
 * @version 1.0
 * @since 31-07-2021
 */

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class FXMLUtil {

    public static FXMLLoader getFXMLLoader(FXInterface fxInterface) {
        return Stream.of(fxInterface)
                .map(FXInterface::toString)
                .map(FXMLUtil.class::getResource)
                .map(FXMLLoader::new)
                .peek(loader -> Optional
                        .of("controls")
                        .map(ResourceBundle::getBundle)
                        .ifPresent(loader::setResources))
                .findFirst()
                .orElse(null);
    }

    public static Pane loadInterface(FXMLLoader loader) {
        Pane pane = null;

        try {
            pane = loader.load();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return pane;
    }

    public static Pane loadInterface(FXInterface fxInterface) {
        return Optional.of(fxInterface)
                .map(FXMLUtil::getFXMLLoader)
                .map(FXMLUtil::loadInterface)
                .orElse(null);

    }
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/util/FXMLUtil.kt
package com.mj.tic.tac.toe.javafx.kotlin.util

import java.util.Optional
import java.util.ResourceBundle
import javafx.fxml.FXMLLoader
import javafx.scene.layout.Pane

/**
 * @author Montaser Jamal
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 25-01-2023
 */

object FXMLUtil {

    fun getFXMLLoader(fxInterface: FXInterface): FXMLLoader {
        return fxInterface
            .toString()
            .let(FXMLUtil.javaClass::getResource)
            .let(::FXMLLoader)
            .also { loader ->
                ResourceBundle.getBundle("controls")?.let(loader::setResources)
            }
    }

    fun loadInterface(loader: FXMLLoader?): Pane? = loader!!.load()

    fun loadInterface(fxInterface: FXInterface): Pane? {
        return Optional.of(fxInterface)
            .map(::getFXMLLoader)
            .map(::loadInterface)
            .orElse(null)
    }
}